### PR TITLE
#7309 add support for building lightweight chassis

### LINKED ISSUE #7309: [Enhancement] Add Gradle support for building the platform with equivalent of  -DadminChassisOnly
### Existing/related issue?

_No response_

### Please describe the new behavior that that will improve Egeria

The egeria.git Maven build has a profile called  `-DadminChassisOnly` that builds the OMAG Server platform so that the registered services are optional.  The platform only loads the services it finds in the classpath.

This option is used for creating cut-down platforms for specific deployment environments.

### Alternatives

The Maven build is retained to continue to provide this option.

### Any Further Information?

_No response_

### Would you be prepared to be assigned this issue to work on?

_No response_

### PR DESCRIPTION
**Pull request by planetf1**
Signed-off-by: Nigel Jones <nigel.l.jones+git@gmail.com>

## Description

Adds support to build a lightweight chassis for gradle builds.

Build by adding `-PadminChassisOnly`, for example
```bash
./gradlew -PadminChassisOnly build -x test -x javadoc
```
to build egeria with a cut-down admin chassis, and skipping test & javadoc.

## Related Issue(s)

Fixes odpi/egeria-docs#675

## Testing

Built with & without flag, and checked content of artifacts ie when set:
```
% jar -tf server-chassis-spring-3.15-SNAPSHOT.jar | grep spring | grep 3.15  
BOOT-INF/lib/admin-services-spring-3.15-SNAPSHOT.jar
BOOT-INF/lib/platform-services-spring-3.15-SNAPSHOT.jar
BOOT-INF/lib/spring-rest-client-connector-3.15-SNAPSHOT.jar
```

## Release Notes & Documentation

Should be added to dojo documentation when happy (@mandy-chessell please indicate if you plan to do this)

## Additional notes

This change replicates what we have with maven (though there we typically set via mvn with `-D` instead of `-P` - the implementation for gradle is more aligned with the normal way of setting gradle properties

In maven there is a conditional build for the connector configuration factory. I've implemented this. However the gradle build doesn't currently build an uber jar for this project, so the change makes no difference..... this condition could be removed. It's only relevant when building an uber jar since the dependencies are runtime only.

For the server chassis
 - gaf wasn't in the gradle build at all. I presume it should be, so added
  - 4 other dependencies looked as if they should have been excluded, but weren't. I have done so for gradle.
  
  Please review. I can consolidate these changes with another commit
  
  NOTE: I have not actually RUN a cut down platform.

### CODE REVIEW

File: open-metadata-implementation/server-chassis/server-chassis-spring/build.gradle [hunk 1]
```diff
@@ -44,59 +44,66 @@ dependencies {
     runtimeOnly 'org.hibernate:hibernate-validator'
     runtimeOnly project(':open-metadata-implementation:platform-services:platform-services-spring')
     runtimeOnly project(':open-metadata-implementation:admin-services:admin-services-spring')
-    runtimeOnly project(':open-metadata-implementation:framework-services:ocf-metadata-management:ocf-metadata-spring')
-    runtimeOnly project(':open-metadata-implementation:repository-services:repository-services-spring')
```
Comment (mandy-chessell) [inline open-metadata-implementation/server-chassis/server-chassis-spring/build.gradle:48]: The truely optional services are the registered services - these are the OMASs (access-services), OMVSs (view-services), OMIS (integration-services) and OMES (engine-services).  Their are loaded into the platform and initialized when their spring module is detected with the component scan.  Excluding their spring module from the loader path ensures none of the associated modules from the service are loaded.  Trying to activate them through a configuration document results in an error.  This ability to completely remove registered services is what we have documented at the moment.

This build option is also the way that new, third-party registered services are introduced.  Their modules are added to the loader path - detected in component scan and loaded as if part of the platform build.  They can alsothen  be configured in a server just like the Egeria provided registered services.

The new (fixed) services listed - the framework services, conformance test services and governance server services - operate differently.  Their server/api modules are always needed by the platform.  If we make their spring modules "compileOnly" and they are not in the loader path, then their REST calls are not available.  Their services can be initialized and started in the platform through configuration documents.  However, the REST API is not available.  This does have some value in that it reduces the attack surface of the platform.

So two choices:
* remove the new fixed services from the list in this PR
* update the documentation on this profile option to explain why/when/how to remove the REST API support for the fixed services.

I favour the second option and volunteer to do the doc updates.  Do you agree?

**Review by mandy-chessell — APPROVED** [review]

File: open-metadata-implementation/server-chassis/server-chassis-spring/build.gradle (hunk 1 above)
Comment (planetf1) [reply open-metadata-implementation/server-chassis/server-chassis-spring/build.gradle:48]: Yes. Can we also check the other appropriate jars are present. This is the current list of jars from egeria that is found in the cut-down chassis, as produced by gradle:
```
admin-services-api-3.15-SNAPSHOT.jar
admin-services-client-3.15-SNAPSHOT.jar
admin-services-registration-3.15-SNAPSHOT.jar
admin-services-server-3.15-SNAPSHOT.jar
admin-services-spring-3.15-SNAPSHOT.jar
asset-lineage-api-3.15-SNAPSHOT.jar
asset-lineage-client-3.15-SNAPSHOT.jar
asset-lineage-topic-connectors-3.15-SNAPSHOT.jar
audit-log-framework-3.15-SNAPSHOT.jar
basic-file-connector-3.15-SNAPSHOT.jar
connector-configuration-factory-3.15-SNAPSHOT.jar
csv-file-connector-3.15-SNAPSHOT.jar
data-engine-api-3.15-SNAPSHOT.jar
data-engine-client-3.15-SNAPSHOT.jar
data-engine-proxy-connector-3.15-SNAPSHOT.jar
data-engine-proxy-services-server-3.15-SNAPSHOT.jar
data-engine-topic-connectors-3.15-SNAPSHOT.jar
engine-host-services-api-3.15-SNAPSHOT.jar
engine-host-services-registration-3.15-SNAPSHOT.jar
engine-host-services-server-3.15-SNAPSHOT.jar
ffdc-services-3.15-SNAPSHOT.jar
gaf-metadata-api-3.15-SNAPSHOT.jar
gaf-metadata-client-3.15-SNAPSHOT.jar
gaf-metadata-server-3.15-SNAPSHOT.jar
generic-handlers-3.15-SNAPSHOT.jar
governance-action-framework-3.15-SNAPSHOT.jar
governance-engine-api-3.15-SNAPSHOT.jar
governance-engine-client-3.15-SNAPSHOT.jar
governance-engine-topic-connectors-3.15-SNAPSHOT.jar
http-helper-3.15-SNAPSHOT.jar
integration-daemon-services-api-3.15-SNAPSHOT.jar
integration-daemon-services-registration-3.15-SNAPSHOT.jar
integration-daemon-services-server-3.15-SNAPSHOT.jar
metadata-security-apis-3.15-SNAPSHOT.jar
metadata-security-connectors-3.15-SNAPSHOT.jar
metadata-security-server-3.15-SNAPSHOT.jar
multi-tenant-3.15-SNAPSHOT.jar
ocf-metadata-api-3.15-SNAPSHOT.jar
ocf-metadata-client-3.15-SNAPSHOT.jar
ocf-metadata-server-3.15-SNAPSHOT.jar
open-connector-framework-3.15-SNAPSHOT.jar
open-discovery-framework-3.15-SNAPSHOT.jar
open-lineage-services-api-3.15-SNAPSHOT.jar
open-lineage-services-server-3.15-SNAPSHOT.jar
open-metadata-archive-file-connector-3.15-SNAPSHOT.jar
open-metadata-conformance-suite-api-3.15-SNAPSHOT.jar
open-metadata-conformance-suite-server-3.15-SNAPSHOT.jar
open-metadata-types-3.15-SNAPSHOT.jar
platform-services-api-3.15-SNAPSHOT.jar
platform-services-server-3.15-SNAPSHOT.jar
platform-services-spring-3.15-SNAPSHOT.jar
repository-handler-3.15-SNAPSHOT.jar
repository-services-apis-3.15-SNAPSHOT.jar
repository-services-archive-utilities-3.15-SNAPSHOT.jar
repository-services-implementation-3.15-SNAPSHOT.jar
rest-client-connectors-api-3.15-SNAPSHOT.jar
rest-client-factory-3.15-SNAPSHOT.jar
spring-rest-client-connector-3.15-SNAPSHOT.jar
```

**Comment by mandy-chessell** [comment]
This is good.  I was surprised by the following libraries but realized they are the client-side libraries pulled in by the governance server services and so it is correct that they are there ...

```
asset-lineage-api-3.15-SNAPSHOT.jar
asset-lineage-client-3.15-SNAPSHOT.jar
asset-lineage-topic-connectors-3.15-SNAPSHOT.jar
data-engine-api-3.15-SNAPSHOT.jar
data-engine-client-3.15-SNAPSHOT.jar
data-engine-topic-connectors-3.15-SNAPSHOT.jar
governance-engine-api-3.15-SNAPSHOT.jar
governance-engine-client-3.15-SNAPSHOT.jar
governance-engine-topic-connectors-3.15-SNAPSHOT.jar
```

**Comment by planetf1** [comment]
> This is good. I was surprised by the following libraries but realized they are the client-side libraries pulled in by the governance server services and so it is correct that they are there ...

Thanks for confirming. I will merge.

We do have options in future to strip more out, but we may also need to bundle some groupings to make it easier to assemble. That can be for another issue if we ever want to.

File: open-metadata-implementation/server-chassis/server-chassis-spring/build.gradle (hunk 1 above)
Comment (planetf1) [reply open-metadata-implementation/server-chassis/server-chassis-spring/build.gradle:48]: has been checked. closing conversation.

**Pull request merged**